src/lib/markbind/src/preprocessors/variablePreprocessor.js

```diff
@@ -0,0 +1,468 @@
+const urlUtils = require('../utils/urls');
+
+const {
+  ATTRIB_CWF,
+  IMPORTED_VARIABLE_PREFIX,
+} = require('../constants');
+
+
+/**
+ * Variable preprocessor to recursively extract and render pages' variables.
+ * Houses a site-wide variable cache.
+ */
+class VariablePreprocessor {
+  constructor(rootPath, baseUrlMap) {
+    /**
+     * Root site path
+     * @type {string}
+     */
+    this.rootPath = rootPath;
+
+    /**
+     * Set of sites' root paths
+     * @type {Set<string>}
+     */
+    this.baseUrlMap = baseUrlMap;
+
+    /**
+     * Map of sites to their variables
+     * @type {Object<string, any>}
+     */
+    this.userDefinedVariablesMap = {};
+
+    /*
+     * Page level
+     */
+
+    /**
+     * Map of files to another object containing its variable names matched to values
+     * @type {Object<string, Object<string, string>>}
+     */
+    this.variableLookup = {};
+
+    /**
+     * Map of files to another object containing its
+     * variable aliases matched to the respective imported file
+     * @type {Object<string, Object<string, string>>}
+     */
+    this.fileAliasesMap = {};
+
+    /**
+     * Set of files that were already processed
+     * @type {Set<string>}
+     */
+    this.processedFiles = new Set();
+  }
+
+
+  /*
+   * --------------------------------------------------
+   * Site level variable storage methods
+   */
+
+
+  /**
+   * Adds a variable for a site with the specified name and value.
+   * @param site path of the site, as reflected by the Site instance's baseUrlMap
+   * @param name of the variable to add
+   * @param value of the variable
+   */
+  addUserDefinedVariable(site, name, value) {
+    this.userDefinedVariablesMap[site] = this.userDefinedVariablesMap[site] || {};
+    this.userDefinedVariablesMap[site][name] = value;
+  }
+
+  /**
+   * Renders the variable in addition to adding it, unlike {@link addUserDefinedVariable}.
+   */
+  addAndRenderUserDefinedVariable(site, name, value) {
+    const renderedVal = njUtil.renderRaw(value, this.userDefinedVariablesMap[site], {}, false);
+    this.addUserDefinedVariable(site, name, renderedVal);
+  }
+
+  /**
+   * Version of {@link addUserDefinedVariable} that adds to all sites.
+   */
+  addUserDefinedVariableForAllSites(name, value) {
+    Object.keys(this.userDefinedVariablesMap)
+      .forEach(base => this.addUserDefinedVariable(base, name, value));
+  }
+
+  resetUserDefinedVariablesMap() {
+    this.userDefinedVariablesMap = {};
+  }
+
+  /*
+   * --------------------------------------------------
+   * Site level variable rendering methods
+   */
+
+
+  /**
+   * Renders the supplied content using only the site variables belonging to the specified base.
+   */
+  renderSiteVariables(siteBase, content) {
+    return njUtil.renderRaw(content, this.userDefinedVariablesMap[siteBase]);
+  }
+
+
+  /*
+   * --------------------------------------------------
+   * Page level variable storage methods
```

Comment:
these will likely be heavily revamped in part 2.

For now, generally, miscallaneous refactors are made to prepare for it ( SLAP, better/more comments, better spacing and reorganisation in methods for logical grouping of functionality, etc )

will point out any outliers